src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,446 @@
+    pallas,
+};
+
+use std::ops::Deref;
+
+impl SinsemillaChip {
+    #[allow(non_snake_case)]
+    #[allow(clippy::type_complexity)]
+    pub(super) fn hash_message(
+        &self,
+        region: &mut Region<'_, pallas::Base>,
+        Q: pallas::Affine,
+        message: &<Self as SinsemillaInstructions<
+            pallas::Affine,
+            { sinsemilla::K },
+            { sinsemilla::C },
+        >>::Message,
+    ) -> Result<(EccPoint, Vec<Vec<CellValue<pallas::Base>>>), Error> {
+        let config = self.config().clone();
+        let mut offset = 0;
+
+        // Get the `x`- and `y`-coordinates of the starting `Q` base.
+        let x_q = *Q.coordinates().unwrap().x();
+        let y_q = *Q.coordinates().unwrap().y();
+
+        // Initialize the accumulator to `Q`.
+        let (mut x_a, mut y_a): (X<pallas::Base>, Y<pallas::Base>) = {
+            // Constrain the initial x_q to equal the x-coordinate of the domain's `Q`.
+            let fixed_x_q =
+                region.assign_fixed(|| "fixed x_q", config.constants, offset, || Ok(x_q))?;
+            let x_q_cell = region.assign_advice(|| "x_q", config.x_a, offset, || Ok(x_q))?;
+            region.constrain_equal(&config.perm, fixed_x_q, x_q_cell)?;
+
+            // This cell gets copied into itself by the first call to `hash_piece` below.
+            let x_a = CellValue::new(x_q_cell, Some(x_q));
+
+            // Constrain the initial x_a, lambda_1, lambda_2, x_p using the fixed y_q
+            // initializer.
+            region.assign_fixed(|| "fixed y_q", config.fixed_y_q, offset, || Ok(y_q))?;
+
+            let y_a = Some(y_q);
+
+            (x_a.into(), y_a.into())
+        };
+
+        let mut zs_sum: Vec<Vec<CellValue<pallas::Base>>> = Vec::new();
+
+        // Hash each piece in the message except the final piece.
+        for piece in message[0..(message.len() - 1)].iter() {
+            // The value of the accumulator after this piece is processed.
+            let (x, y, zs) = self.hash_piece(region, offset, piece, x_a, y_a)?;
+
+            // Since each message word takes one row to process, we increase
+            // the offset by `piece.num_words` on each iteration.
+            offset += piece.num_words();
+
+            // Update the accumulator to the latest value.
+            x_a = x;
+            y_a = y;
+            zs_sum.push(zs);
+        }
+
+        // Hash the final message piece.
+        let y_a = {
+            let piece = &message[message.len() - 1];
+            // The value of the accumulator after this piece is processed.
+            let (x, y, zs) = self.hash_piece(region, offset, piece, x_a, y_a)?;
+
+            // Since each message word takes one row to process, we increase
+            // the offset by `piece.num_words` on each iteration.
+            offset += piece.num_words();
+
+            // Update the accumulator to the latest value.
+            x_a = x;
+            y_a = y;
+            zs_sum.push(zs);
+
```

Comment:
This is identical to the main loop, and AFAICT none of the temporaries generated in this section are used below, so we should just do this in the main loop (but keep the rest of this section for assigning the final `y_a`:
```suggestion
```